### ISSUE TITLE
Authorization fails when using an ssh-agent

### ISSUE THREAD

**Issue by eminence** [Eugeny/russh#189]
I'm trying to use a key loaded into an ssh-agent for authentication.  This is not working.  However, the key works when loaded from disk using `load_secret_key`.

I generated a new RSA key with `ssh-keygen`.

Here's the fingerprint of the key:

```
❯ ssh-keygen -lf lxd_deleteme
2048 SHA256:dd8u94XvU0tJfdPiL/YpU4G0SHNBDJBvrP/6T8HVLO8 lxd@lin-lxd01sim (RSA)
```

I've also loaded it into an ssh-agent.  The ssh-agent reports the identical fingerprint:

```
❯ ssh-add -l
2048 SHA256:dd8u94XvU0tJfdPiL/YpU4G0SHNBDJBvrP/6T8HVLO8 ./lxd_deleteme (RSA)
```

I'm connecting to the ssh-agent like this:

```rust
let mut agent = russh_keys::agent::client::AgentClient::connect_env().await.unwrap();
let mut identities = agent.request_identities().await.unwrap();
assert_eq!(identities.len(), 1); 
let id = identities.pop().unwrap();
println!("Key from ssh-agent: {} {}", id.name(), id.fingerprint());
```

And I'm loading the key from disk like this:

```rust
 let key_pair = load_secret_key("lxd_deleteme", None).unwrap();
```

I first try authentication using `authenticate_future` and then fallback to `authenticate_publickey`.

Here's what the full code looks like:

```rust
#[tokio::main]
async fn main() {
    env_logger::init();
    let config = russh::client::Config::default();
    let config = Arc::new(config);
    let sh = Client {};
    let host = "127.0.0.1";

    let mut agent = russh_keys::agent::client::AgentClient::connect_env().await.unwrap();
    let mut identities = agent.request_identities().await.unwrap();

    assert_eq!(identities.len(), 1);
    let id = identities.pop().unwrap();
    println!("Key from ssh-agent: {} {}", id.name(), id.fingerprint());

    let key_pair = load_secret_key("lxd_deleteme", None).unwrap();
    if let russh_keys::key::KeyPair::RSA { key, hash } = &key_pair {
        println!("Key loaded from disk: hash={hash:?}");
    }

    let mut session = russh::client::connect(config, (host, 22), sh).await.unwrap();
    println!("Connected!");

    // first try with agent auth
    let (_, auth_res) = session.authenticate_future("root", id, agent).await;
    let auth_res = auth_res.unwrap();
    dbg!(auth_res);
    if !auth_res {
        println!("Auth failed using ssh-agent, trying key from disk");
    }
    let auth_res = session
        .authenticate_publickey("root", Arc::new(key_pair))
        .await
        .unwrap();
    dbg!(auth_res);

    println!("=== auth: {}", auth_res);
}
```

And here's the output:

```
Key from ssh-agent: rsa-sha2-512 dd8u94XvU0tJfdPiL/YpU4G0SHNBDJBvrP/6T8HVLO8
Key loaded from disk: hash=SHA2_256
check_server_key: e17SK8zBaPtvQBpIcnt534kbxe2Mhi0tr5XOc/FUR/E
Connected!
src/main.rs:52] auth_res = false
Auth failed using ssh-agent, trying key from disk
src/main.rs:60] auth_res = true
=== auth: true 
```

Any suggestions about what's going wrong, or how to debug this?

## Version info

I'm using russh e5f12248359e9b2a3aed3420c02fbfbb6e4dca84

**Comment by eminence** [comment]
I'm not familiar with russh's code, but I'm hoping to find some time to dig into this issue in the coming weeks.  If anyone has any hints about how to debug this, please let me know

**Comment by Eugeny** [comment]
Running with `RUST_LOG=debug` should give a detailed trace of what's going on.

**Comment by eminence** [comment]
I tried that, but unfortunately the result wasn't that enlightening to me.  Would it be useful if I uploaded the debug traces to this issue, though?

**Comment by Eugeny** [comment]
Definitely!

**Comment by tesujimath** [comment]
I hit this problem, and found I had to use an ED25519 key, since it didn't work with my default RSA key.

```
        let mut agent = AgentClient::connect_env().await?;
        let identities = agent.request_identities().await?;
        // find a compatible identity
        let id = identities
            .into_iter()
            .find(|id| id.name() == "ssh-ed25519")
            .ok_or(anyhow!("ssh agent has no ed25519 identities"))?;
        debug!("using identity {} {}", id.name(), id.fingerprint());
```

Hope that helps you.

**Comment by Eugeny** [comment]
@tesujimath if you're on v0.43 or earlier, you need to enable the `openssl` crate feature to use RSA keys.

**Comment by eminence** [comment]
This issue is actually fixed now. 

It was fixed via 2bfe4260f3928279bc95f27257561eb75b40b2e5 in #294
